### PR TITLE
update parseDeliveredAmount for cases where delivered_amount is not p…

### PR DESCRIPTION
**Pull request by shekenahglory**
following logic used in rippled for handling delivered amount

### CODE REVIEW

File: src/ledger/parse/utils.js [hunk 1]
```diff
@@ -58,13 +58,32 @@ function isPartialPayment(tx) {
...
+    } else if (tx.meta.DeliveredAmount) {
+      deliveredAmount = parseAmount(tx.meta.DeliveredAmount);
+
+    // no partial payment flag
     } else if (tx.Amount && !isPartialPayment(tx)) {
       deliveredAmount = parseAmount(tx.Amount);
+
+    // partial payment after 4594094 without DeliveredAmount
+    } else if (tx.Amount && tx.ledger_index > 4594094) {
+      deliveredAmount = parseAmount(tx.Amount);
```
Comment (lumberj) [inline src/ledger/parse/utils.js]: If it's a partial payment, doesn't that mean the delivered amount is _not_ equal to `tx.Amount`?

Comment (shekenahglory) [inline src/ledger/parse/utils.js]: nope, partial payment flag means it can be successful even if its not the
full amount, but `DeliveredAmount` only shows up if it was less than the
full amount - so after 4594094 a tx with partial payment flag without
DeliveredAmount means it delivered the full amount.

**Comment by lumberj** [comment]
LGTM

File: docs/index.md [hunk 2]
```diff
@@ -839,6 +839,7 @@ outcome | object | The outcome of the transaction (what effects it had).
 *outcome.* ledgerVersion | integer | The ledger version that the transaction was validated in.
 *outcome.* indexInLedger | integer | The ordering index of the transaction in the ledger.
 *outcome.* deliveredAmount | [amount](#amount) | *Optional* For payment transactions, it is impossible to reliably compute the actual delivered amount from the balanceChanges due to fixed precision. If the payment is not a partial payment and the transaction succeeded, the deliveredAmount should always be considered to be the amount specified in the transaction.
+*outcome.* deliveredAmount | string | *Optional* The delivered amount is reported as 'unavailable' for partial payment transactions before 2014
```
Comment (clark800) [inline docs/index.md]: `deliveredAmount` is already an optional field, so why do you have to say that it is "unavailable" as opposed to just not including it?

File: src/ledger/parse/utils.js [hunk 3]
```diff
@@ -58,13 +58,32 @@ function isPartialPayment(tx) {
...
-  // TODO: Workaround for existing rippled bug where delivered_amount may not be
-  // provided for account_tx
-  if (tx.TransactionType === 'Payment') {
-    if (tx.meta.delivered_amount) {
+  if (tx.TransactionType === 'Payment' &&
+      tx.meta.TransactionResult === 'tesSUCCESS') {
+
+    if (tx.meta.delivered_amount &&
+        tx.meta.delivered_amount === 'unavailable') {
+      deliveredAmount = 'unavailable';
```
Comment (clark800) [inline src/ledger/parse/utils.js]: How about just returning in each branch instead of assigning to a variable?

File: src/ledger/parse/ledger.js [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 const {removeUndefined, rippleTimeToISO8601} = require('./utils');
 const parseTransaction = require('./transaction');
 import type {GetLedger} from '../types.js';
 
 function parseTransactionWrapper(ledgerVersion, tx) {
-  const transaction = _.assign({}, _.omit(tx, 'metaData'),
-    {meta: tx.metaData});
+  const transaction = _.assign({}, _.omit(tx, 'metaData'), {
+    meta: tx.metaData,
+    ledger_index: ledgerVersion
```
Comment (clark800) [inline src/ledger/parse/ledger.js]: Why is this necessary?

File: src/ledger/parse/utils.js (hunk 1 above)
Comment (clark800) [inline src/ledger/parse/utils.js]: I think we need a comment for this, along with an explanation for what changed in ledger version 4594094

File: src/ledger/parse/utils.js (hunk 3 above)
Comment (clark800) [inline src/ledger/parse/utils.js]: You could also break the else-if chain into just a sequence of if-returns which would make the comments less awkward

File: docs/index.md (hunk 2 above)
Comment (shekenahglory) [inline docs/index.md]: It has to be optional because it can only be present for successful Payment transactions, part of the reason they created the "delivered_amount" decorator because "DeliveredAmount" was not always present, so it was potentially confusing to understand exactly how much was transferred....this way it would always be present when *some* amount is transferred, even if it can't explictly state what that amount is.

Comment (clark800) [inline docs/index.md]: > this way it would always be present when some amount is transferred, even if it can't explictly state what that amount is

What is the advantage of that if it's going to require a special case anyways?

Comment (shekenahglory) [inline docs/index.md]: I don't think it matters much either way, since it will only be missing/unavailable on a few partialPayment transactions pre 2014 (once rippled team corrects their logic)

File: src/ledger/parse/ledger.js (hunk 4 above)
Comment (shekenahglory) [inline src/ledger/parse/ledger.js]: its another field that rippled includes in a 'tx' request, but not in a 'ledger' request - I need it at least for the ledger_index check in parseDeliveredAmount, I didn't check to see if it was expected and used elsewhere

File: docs/index.md (hunk 2 above)
Comment (clark800) [inline docs/index.md]: Ok, I think it's better to just omit it then, it will be less likely to produce bugs in the client code

**Comment by clark800** [comment]
LGTM
